src/components/common/IdContainer.ts

```diff
@@ -0,0 +1,32 @@
+import _ from 'lodash/fp'
+import { useMemo, useState } from 'react'
+
+/**
+ * DEPRECATED - should switch to useUniqueIdFn pattern
+ */
+export const IdContainer = ({ children }) => {
+  const [id] = useState(() => _.uniqueId('element-'))
+  return children(id)
+}
+
+export type UniqueIdFn = (prefix: string) => string
+
+/**
+ * returns a helper function that accepts a prefix to append to an id that is unique
+ * for the consuming component.
+ * @example
+ * // inside a react component:
+ * const unique = useUniqueIdFn()
+ * //...
+ * div({ id: unique('my-thing-one') }, [...],
+ * div({ id: unique('my-thing-two') }, [...]
+ * // will produce ids:
+ * // my-thing-one-123UniqueToken123
+ * // my-thing-two-123UniqueToken123
+ */
+export const useUniqueIdFn = (): UniqueIdFn => {
+  const uniqueId: string = useMemo(() => _.uniqueId(''), [])
+  return ((prefix: string): string => {
+    return `${prefix}-${uniqueId}`
+  })
```

Comment:
I had the same quesiton/debate in my head.... 
- getMyId fn is a little easier on they keystrokes (no ` and $ keys required)
- but the `prefix-${myId}` pattern is more transparent and simpler on the return...

The tie-breaker for me was to more closely model the pattern this was replacing... which worked the prefix into the flow of things... but make the prefix authorable.

I think that, and unique('button-a') seemed to read nicely to me.

Reply:
I think the most common use case for IdContainer is linking form elements and their labels.

I think I'd rather write:
```js
const id = useUniqueId()

...
label({ htmlFor: id }, ...),
h(Input, { id }),
```

Than
```js
const getId = useUniqueIdFn()
const id = getId()

...
label({ htmlFor: id }, ...),
h(Input, { id }),
```

or

```js
const getId = useUniqueIdFn()

...
label({ htmlFor: getId() }, ...),
h(Input, { id: getId() }),
```
 
And having the hook return an ID instead of a function leaves it up to the consumer how/if they want to construct derived IDs, like adding a prefix or a suffix.

Reply:
I strongly prefer the 1st of these 3 options. I would expect every time that you call `useUniqueId()`, you get a unique ID. Otherwise I really don't understand the point of the method.

But again, what I really want to do most of the time is simply nest the input inside of the label, as then you don't need the ids at all. Is there a way to do this with our way of writing React?

```
<label>
  First name:
  <input type="text" name="firstname">
</label>

Reply:
> Is there a way to do this with our way of writing React?

Yes.

```js
label([
  'First name:',
  h(TextInput, ...)
])
```

Reply:
Ah right, that looked odd to me but obviously….

I would suggest we encourage people to write more in this style, as then ids are needed far less often. Assuming of course that we don't emit a false alarm in this case via the `id` assertion.